File: apps/matriz-esp32/src/ConfigManager.cpp
#include "ConfigManager.h"
#include <LittleFS.h>
#include <ArduinoJson.h>

static const char* CONFIG_PATH = "/config_matriz.json";
static MatrizConfig config;

bool ConfigManager::begin() {
	if (!LittleFS.begin(true)) {
		Serial.println("LittleFS Mount Failed");
		return false;
	}

	File file = LittleFS.open(CONFIG_PATH, "r");
	if (!file) {
		return true;
	}

	String content = file.readString();
	file.close();

	JsonDocument doc;
	DeserializationError error = deserializeJson(doc, content);
	if (error) {
		return true;
	}

	config.user = doc["user"] | "";
	config.pass = doc["pass"] | "";
	config.polling_ms = doc["polling_ms"] | 0;

	JsonArray filiais = doc["filiais"].as<JsonArray>();
	for (JsonObject f : filiais) {
		FilialInfo info;
		info.name = f["name"] | "";
		info.ip = f["ip"] | "";
		info.port = f["port"] | 51000;
		config.filiais.push_back(info);
	}

	return true;
}

String ConfigManager::getConfigJson() {
	File file = LittleFS.open(CONFIG_PATH, "r");
	if (!file) {
		return "{}";
	}

	String content = file.readString();
	file.close();
	return content;
}

bool ConfigManager::saveConfig(const String& body) {
	JsonDocument doc;
	DeserializationError error = deserializeJson(doc, body);
	if (error) {
		Serial.println("JSON parse error");
		return false;
	}

	if (!doc["user"].is<const char*>() || !doc["pass"].is<const char*>()) {
		return false;
	}

	File file = LittleFS.open(CONFIG_PATH, "w");
	if (!file) {
		Serial.println("Failed to open config for writing");
		return false;
	}

	if (serializeJson(doc, file) == 0) {
		file.close();
		return false;
	}

	file.close();

	config.user = doc["user"] | config.user;
	config.pass = doc["pass"] | config.pass;
	config.polling_ms = doc["polling_ms"] | config.polling_ms;

	JsonArray filiais = doc["filiais"].as<JsonArray>();
	if (!filiais.isNull()) {
		config.filiais.clear();
		for (JsonObject f : filiais) {
			FilialInfo info;
			info.name = f["name"] | "";
			info.ip = f["ip"] | "";
			info.port = f["port"] | 51000;
			config.filiais.push_back(info);
		}
	}

	return true;
}

MatrizConfig& ConfigManager::getConfig() {
	return config;
}

File: apps/udp-iot/applications/matriz-java/src/matriz/ConfigManager.java
package matriz;
import lib.logging.Logger;

import shared.Json;
import shared.Json.JsonArray;
import shared.Json.JsonObject;

import java.io.IOException;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;

/**
 * Manages the Matriz configuration.
 *
 * <p>Replaces {@code matriz-esp32/ConfigManager.cpp}.
 * Loads from and saves to a JSON config file.
 * Thread-safe: reads/writes are synchronised on the config object.
 */
public class ConfigManager {

    private static final Logger logger = Logger.getLogger(ConfigManager.class);

    private volatile MatrizConfig config;
    private String configPath;

    /**
     * Load configuration from a JSON file.
     * @param path filesystem path to config_matriz.json
     * @return true if loaded successfully
     */
    public synchronized boolean load(String path) {
        this.configPath = path;
        try {
            this.config = readConfigFromDisk();
            return true;
        } catch (IOException e) {
            logger.error("ConfigManager: IO error: {}", e.getMessage());
            return false;
        } catch (RuntimeException e) {
            logger.error("ConfigManager: Parse error: {}", e.getMessage());
            return false;
        }
    }

    /**
     * Save current configuration to the JSON file.
     * @param jsonContent the full JSON string to write
     * @return true if saved successfully
     */
    public synchronized boolean save(String jsonContent) {
        try {
            // Validate that it parses and has required fields
            JsonObject obj = Json.parseObject(jsonContent);
            if (!obj.has("user") || !obj.has("pass")) {
                logger.warn("ConfigManager: Refusing to save — missing user/pass");
                return false;
            }

            MatrizConfig parsed = parseConfig(obj);
            writeConfigToDisk(jsonContent);
            this.config = readConfigFromDisk();
            logger.info("ConfigManager: Configuration saved ({} filiais)", parsed.filiais().size());
            return true;
        } catch (IOException e) {
            logger.error("ConfigManager: IO error on save: {}", e.getMessage());
            return false;
        } catch (RuntimeException e) {
            logger.error("ConfigManager: Invalid JSON on save: {}", e.getMessage());
            return false;
        }
    }

    /** Return the current config as a JSON string (for REST API). */
    public synchronized String getConfigJson() {
        try {
            return Files.readString(Path.of(configPath));
        } catch (IOException e) {
            throw new RuntimeException("ConfigManager: Could not read config JSON", e);
        }
    }

    /** Return the current parsed config object. */
    public synchronized MatrizConfig getConfig() {
        try {
            this.config = readConfigFromDisk();
            return config;
        } catch (IOException e) {
            throw new RuntimeException("ConfigManager: Could not refresh config from disk", e);
        }
    }

    /**
     * Find a filial's port by IP address.
     * @return port number, or 51000 if not found
     */
    public synchronized int findPortByIp(String ip) {
        for (FilialInfo f : getConfig().filiais()) {
            if (f.ip().equals(ip)) return f.port();
        }
        return 51000;
    }

    private MatrizConfig readConfigFromDisk() throws IOException {
        String content = Files.readString(Path.of(configPath));
        JsonObject obj = Json.parseObject(content);
        return parseConfig(obj);
    }

    private MatrizConfig parseConfig(JsonObject obj) {
        String user = obj.getString("user", "admin");
        String pass = obj.getString("pass", "admin");
        int pollingMs = resolvePollingMs(obj);

        List<FilialInfo> filiais = new ArrayList<>();
        if (obj.has("filiais")) {
            JsonArray arr = obj.getArray("filiais");
            for (int i = 0; i < arr.size(); i++) {
                JsonObject f = arr.getObject(i);
                String name = f.getString("name", "Filial " + (i + 1));
                String ip   = f.getString("ip", "127.0.0.1");
                int port    = f.getInt("port", 51000);
                filiais.add(new FilialInfo(name, ip, port));
            }
        }

        return new MatrizConfig(user, pass, pollingMs, filiais);
    }

    private int resolvePollingMs(JsonObject obj) {
        if (obj.has("pollingMs")) {
            return obj.getInt("pollingMs", 0);
        }
        return obj.getInt("polling_ms", 0);
    }

    private void writeConfigToDisk(String jsonContent) throws IOException {
        Path target = Path.of(configPath);
        Path parent = target.toAbsolutePath().getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }

        Path tempFile = Files.createTempFile(parent, target.getFileName().toString(), ".tmp");
        try {
            Files.writeString(tempFile, jsonContent);
            try {
                Files.move(
                    tempFile,
                    target,
                    StandardCopyOption.REPLACE_EXISTING,
                    StandardCopyOption.ATOMIC_MOVE
                );
            } catch (AtomicMoveNotSupportedException e) {
                Files.move(tempFile, target, StandardCopyOption.REPLACE_EXISTING);
            }
        } finally {
            Files.deleteIfExists(tempFile);
        }
    }

    /** Configuration data record. */
    public record MatrizConfig(
        String user,
        String pass,
        int pollingMs,
        List<FilialInfo> filiais
    ) {}
}
